consensus/polybft/proposer_calculator.go

```diff
@@ -0,0 +1,499 @@
+		if val, exists := addressOldValidatorMap[x.Address]; exists {
+			priority = val.ProposerPriority // + int64(val.Metadata.VotingPower) - int64(x.VotingPower)
+		}
+
+		newValidatorsCalcProposer[i] = &PrioritizedValidator{
+			Metadata:         x,
+			ProposerPriority: priority,
+		}
+	}
+
+	// TODO: centering
+
+	snapshot.Validators = newValidatorsCalcProposer
+}
+
+func incrementProposerPriority(snapshot *ProposerSnapshot, totalVotingPower int64) (*PrioritizedValidator, error) {
+	for _, val := range snapshot.Validators {
+		// Check for overflow for sum.
+		newPrio := safeAddClip(val.ProposerPriority, int64(val.Metadata.VotingPower))
+		val.ProposerPriority = newPrio
+	}
+	// Decrement the validator with most ProposerPriority.
+	mostest, err := getValWithMostPriority(snapshot)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get validator with most priority: %w", err)
+	}
+
+	mostest.ProposerPriority = safeSubClip(mostest.ProposerPriority, totalVotingPower)
+
+	return mostest, nil
+}
+
+func updateWithChangeSet(snapshot *ProposerSnapshot, totalVotingPower int64) error {
+	if err := rescalePriorities(snapshot, totalVotingPower); err != nil {
+		return fmt.Errorf("cannot rescale priorities: %w", err)
+	}
+
+	if err := shiftByAvgProposerPriority(snapshot); err != nil {
+		return fmt.Errorf("cannot shift proposer priorities: %w", err)
+	}
+
+	return nil
+}
+
+func shiftByAvgProposerPriority(snapshot *ProposerSnapshot) error {
+	avgProposerPriority, err := computeAvgProposerPriority(snapshot)
+	if err != nil {
+		return fmt.Errorf("cannot compute proposer priority: %w", err)
+	}
+
+	for _, val := range snapshot.Validators {
+		val.ProposerPriority = safeSubClip(val.ProposerPriority, avgProposerPriority)
+	}
+
+	return nil
+}
+
+func getValWithMostPriority(snapshot *ProposerSnapshot) (result *PrioritizedValidator, err error) {
+	if len(snapshot.Validators) == 0 {
+		return nil, fmt.Errorf("validators cannot be nil or empty")
+	}
+
+	for _, curr := range snapshot.Validators {
+		// pick curr as result if it has greater priority
+		// or if it has same priority but "smaller" address
+		if isBetterProposer(curr, result) {
+			result = curr
+		}
+	}
+
+	return result, nil
+}
+
+func computeAvgProposerPriority(snapshot *ProposerSnapshot) (int64, error) {
+	if len(snapshot.Validators) == 0 {
+		return 0, fmt.Errorf("validator set cannot be nul or empty")
+	}
+
+	n := int64(len(snapshot.Validators))
+	sum := big.NewInt(0)
+
+	for _, val := range snapshot.Validators {
+		sum.Add(sum, big.NewInt(val.ProposerPriority))
+	}
+
+	avg := sum.Div(sum, big.NewInt(n))
+	if avg.IsInt64() {
+		return avg.Int64(), nil
+	}
+
+	return 0, fmt.Errorf("cannot represent avg ProposerPriority as an int64 %v", avg)
+}
+
+// rescalePriorities rescales the priorities such that the distance between the
+// maximum and minimum is smaller than `diffMax`.
+func rescalePriorities(snapshot *ProposerSnapshot, totalVotingPower int64) error {
```

Comment:
```suggestion
func (snapshot *ProposerSnapshot) rescalePriorities(totalVotingPower int64) error {
```
If possible, lets not follow a function approach and create internal methods for the type.

Reply:
Can this be a new task? It will be easier after we finish the whole feature.

Reply:
Please add to Jira.

Reply:
Here is the task: https://polygon.atlassian.net/browse/EVM-330